module/zfs/spa_errlog.c

```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
 	mutex_exit(&spa->spa_errlist_lock);
 }
 
+#ifdef _KERNEL
+static int
+find_block_txg(dsl_dataset_t *ds, zbookmark_err_phys_t *zep,
+    uint64_t *birth_txg)
+{
+	objset_t *os;
+	int error = dmu_objset_from_ds(ds, &os);
+	if (error != 0)
+		return (error);
+
+	dnode_t *dn;
+	blkptr_t bp;
+
+	error = dnode_hold(os, zep->zb_object, FTAG, &dn);
+	if (error != 0)
+		return (error);
+
+	rw_enter(&dn->dn_struct_rwlock, RW_READER);
+	error = dbuf_dnode_findbp(dn, zep->zb_level, zep->zb_blkid, &bp, NULL,
+	    NULL);
+
+	if (error == 0 && BP_IS_HOLE(&bp))
+		error = SET_ERROR(ENOENT);
+
+	*birth_txg = bp.blk_birth;
+	rw_exit(&dn->dn_struct_rwlock);
+	dnode_rele(dn, FTAG);
+	return (error);
+}
+
+/*
+ * This function serves a double role. If only_count is true, it returns
+ * how many times an error block belonging to this filesystem is referenced by
+ * snapshots or clones. If only_count is false, each time the error block is
+ * referenced by a snapshot or clone, it returns a pointer to the bookmark of
+ * the error block back to process_error_log().
+ */
+static int
+check_filesystem(spa_t *spa, uint64_t fs, zbookmark_err_phys_t *zep,
+    uint64_t *count, void *addr, boolean_t only_count)
+{
+	dsl_dataset_t *ds;
+	dsl_pool_t *dp = spa->spa_dsl_pool;
+
+	int error = dsl_dataset_hold_obj(dp, fs, FTAG, &ds);
+	if (error != 0)
+		return (error);
+
+	uint64_t latest_txg;
+	uint64_t txg_to_consider = spa->spa_syncing_txg;
+	boolean_t check_snapshot = B_TRUE;
+	error = find_block_txg(ds, zep, &latest_txg);
+	if (error == 0) {
+		if (zep->zb_birth < latest_txg) {
+			txg_to_consider = latest_txg;
+		} else {
+			/* Block neither free nor rewritten. */
+			if (!only_count) {
+				zbookmark_phys_t zb;
+				zep_to_zb(fs, zep, &zb);
+				if (copyout(&zb, (char *)addr + (*count - 1)
+				    * sizeof (zbookmark_phys_t),
+				    sizeof (zbookmark_phys_t)) != 0) {
+					dsl_dataset_rele(ds, FTAG);
+					return (SET_ERROR(EFAULT));
+				}
+				(*count)--;
+			} else {
+				(*count)++;
+			}
+			check_snapshot = B_FALSE;
+		}
+	}
+
+	/* How many snapshots reference this block. */
+	uint64_t snap_count;
+	error = zap_count(spa->spa_meta_objset,
+	    dsl_dataset_phys(ds)->ds_snapnames_zapobj, &snap_count);
+	if (error != 0) {
+		dsl_dataset_rele(ds, FTAG);
+		return (error);
+	}
+
+	if (snap_count == 0) {
+		/* File system has no snapshot. */
+		dsl_dataset_rele(ds, FTAG);
+		return (0);
+	}
+
+	uint64_t *snap_obj_array = kmem_alloc(snap_count * sizeof (uint64_t),
+	    KM_SLEEP);
+
+	int aff_snap_count = 0;
+	uint64_t snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	uint64_t snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
```

Comment:
Since check_filesystem() is passed the *oldest* affected snapshot, I think that this should be walking *forward*, but it's walking backwards, so I'm confused.  Can you explain how this all works?

Reply:
See my comments above.